### ISSUE TITLE
[Regression] Batman Arkham City - Rendering issue 

### ISSUE THREAD

**Issue by Jonathan44062** [RPCS3/rpcs3#15225]
### Quick summary

Without realizing it, Batman Arkham City graphics are now corrupted, even when the Write Depth Buffers is enabled

### Details

Working - 0.0.25-14483 
https://drive.google.com/file/d/11ytzgGQv1qyQ6Zg3uijmbDf9RSaa-Pdt/view?usp=sharing

Regression - 0.0.25-14488 https://github.com/RPCS3/rpcs3/pull/13077
https://drive.google.com/file/d/1Ur50DcyI3AaAIcAxxhd-7ZA8RKb4qNtJ/view?usp=sharing

### Build with regression

0.0.25-14488

### Attach two log files

Working
[BAC Works.log.gz](https://github.com/RPCS3/rpcs3/files/14339625/BAC.Works.log.gz)

Broken
[BAC Broken.log](https://github.com/RPCS3/rpcs3/files/14339673/BAC.Broken.log)

### Attach capture files for visual issues

https://www.mediafire.com/file/ldokztjp5xq2ll0/BLUS30978_20240219231033_capture.rrc/file

### System configuration

Windows 10 22H2
Intel Core i7-8700
15.93 GiB RAM
GPU: NVIDIA GeForce GTX 1070 Ti (551.52)

### Other details

I can share my save file to test better the issue I posted, this is the GOTY Version
https://www.mediafire.com/file/92t7rx0abxidk08/BLUS30978-SAVE0.rar/file

**Comment by kd-11** [comment]
Why is it never the easy PRs? 😞

**Comment by Jonathan44062** [comment]
Still an issue.

**Comment by digant73** [comment]
that broken PR is small and the code is still the same in recent builds. So I simply changed only one line of code on the recent build just to make a test.

@Jonathan44062 could you verify if the attached `rpcs3.exe` fixes the issue? 

eventually the changed code is at line 1658 on `texture_cache_utils.h`:

`	if (/*!should_flush() ||*/ get_confirmed_range().inside(flush_exclusions))
`

just commented out a potential wrong evaluated condition not present on working build 0.0.25-14483

[rpcs3.zip](https://github.com/user-attachments/files/20128641/rpcs3.zip)

**Comment by Jonathan44062** [comment]
Your build fixes the broken rendering in Arkham City, same applies for Akham Orings, I tested Arkham Asylum and everything looks fine

**Comment by digant73** [comment]
thanks for checking. Of course, It's up to kd-11 to investigate on that condition and `should_flush()` function and provide the fix. Hopefully other bug reports are related to this bug

**Comment by kd-11** [comment]
Hmm, this one is tricky, I'll take a look at that function.
The optimization is present to skip flushing data when nothing has changed, looks like we may have missed an edge case.

**Comment by digant73** [comment]
In the meanwhile I'm using that temporary fix on other games. Apparently no regression on visual issues and performance

**Comment by kd-11** [comment]
@digant73 @Jonathan44062 Is anyone able to reproduce the issue using the demo version of this game? I don't see any issues so far.
In fact it was not clear from the start what the visual issue is. Is it the geometry pop-in/culling? I am using default settings + WDB only.

**Comment by digant73** [comment]
From the posted video the issue was disappearing and re-appearing textures when moving the camera.
I will see the demo

**Comment by digant73** [comment]
Using only WDB, yes I see the missing or grey textures just at the beginning of the demo. Simply rotating the camera. I will try to upload a video

**Comment by kd-11** [comment]
I cannot cause this symptom to appear, but maybe it's just tied to the user's CPU. Neither of my GPUs can reproduce the issue.

**Comment by digant73** [comment]
I had a 7800x3d and now 9950x3d. same issue

**Comment by kd-11** [comment]
> I had a 7800x3d and now 9950x3d

I have a 5900X so maybe my slower CPU is hiding the symptoms somehow. That temp fix doesn't really mean much in this case, it's just hiding a complex side-effect by bypassing something important which is tag synchronization.
Try commenting out these lines and see if the issue goes away: https://github.com/RPCS3/rpcs3/blob/master/rpcs3/Emu/RSX/Common/surface_store.h#L1255-L1258

**Comment by kd-11** [comment]
I was able to make the issue appear only when uncapping fps and using high vblank count so that my fps exceeds the stock 30fps.
Is this the observation everyone else has?

**Comment by digant73** [comment]
no I had the issue even without uncapping fps (using only WDC).
Issue is present even with your suggested piece of code commented out

**Comment by digant73** [comment]
here another video where the issue is very clear

https://mega.nz/file/fv4RwSRS#ytVBsZF49Kh9FDkk232b2t_0AemQnIcBNHUxfH6AKK4

**Comment by kd-11** [comment]
I figured out what's happening here:
```
Lock 0 -> Fault 0 -> Unlock+Flush 0 -> Write 1 -> Tag 1 -> Lock 2 -> Read 2 (No tag!!!) -> Fault 2(No flush)
```
Nasty timeline. A fault must be getting triggered from RSX itself between draw call begin and draw call end. Quite the design flaw, I'll just delay RTT setup.

**Comment by digant73** [comment]
ok.

Below another video:

https://mega.nz/file/WjpGgQjA#PZjL7SLw2t3tTYYgTM_vmQ5GN6CO6849gjWcNRv-Rj4

**Comment by kd-11** [comment]
https://github.com/RPCS3/rpcs3/pull/17203 fixes this issue.

**Comment by Bobble80** [comment]
> > [#17203](https://github.com/RPCS3/rpcs3/pull/17203) fixes this issue.
> 
> good. I will test it this night and make some no regression tests. Is it possible that this issue was present even before [#13077](https://github.com/RPCS3/rpcs3/pull/13077)? so the bug is responsible for other flickering textures reported on older bug reports such as in R6C [#13907](https://github.com/RPCS3/rpcs3/issues/13907), [#16873](https://github.com/RPCS3/rpcs3/issues/16873) etc.?

No, unfortunately didn’t fix the flickering graphics issue in Splatterhouse. 

https://github.com/RPCS3/rpcs3/issues/16873

I thought it might have, as it was a similar looking issue as the one in Batman Arkham.

**Comment by digant73** [comment]
ok. thanks for checking
